# 기계학습 (60점)

In [1]:
from IPython.display import display, HTML
display(HTML("<style>:root {--jp-notebook-max-width: 3000px;}</style>"))
display(HTML("<style>.container { width:100% !important; }</style>"))

### 출생데이터
- 출처 : https://mdis.kostat.go.kr/dwnlSvc/ofrSurvSearch.do?curMenuNo=UI_POR_P9240
- 자료 이용 주의사항 (국가통계 자료제공 규정 제 17조 제 1항)     
  ○ 개별 자료에 의거 알게 된 사항에 대한 제공 및 누설 금지    
  ○ 자료 이용의 정확한 목적 명시 및 작성 목적 이외의 사용 금지     
  ○ 제공 자료의 활용이 끝난 후 즉시 파기    
  ○ 자료의 무단 공유 · 복제 금지     
  ○ 올바른 분석기법 사용 및 통계적 오차를 적정수준으로 유지하도록 노력     

- 본 모의고사는 MDIS 담당부서 문의 후 자료 이용 동의 받고 사용

데이터 경로 
- ./data/
    - 2019.csv ~ 2022.csv : 년도별 출생관련 데이터
    - index.csv : 데이터 컬럼 상세
        - 항목명 : 컬럼
            - 항목명 컬럼의 결측치는 값이 존재하는 상위 데이터와 일치한다. 
            - ex. 1행의 항목명은 '출생자주소지_행정구역시도코드' , 18행의 항목명은 성별코드
        - 코드 : 데이터
        - 코드의미 및 단위 : 코드의 의미

In [2]:
import pandas as pd
pd.set_option('display.max_columns',100)


df =pd.read_csv('./data/2019.csv')
display(df.head())
idx = pd.read_csv('./data/index.csv')
idx.head()

연도  신고연도  신고월  신고일  출생자주소지_행정구역시도코드  성별코드 결혼중외의자녀여부코드  출생연도  출생월  출생장소코드  \
0  2019  2019    1    1               31     1           Y  2019    1       2   
1  2019  2019    1    1               33     1           9  2019    1       2   
2  2019  2019    1    2               11     1           Y  2019    1       2   
3  2019  2019    1    2               11     1           Y  2019    1       2   
4  2019  2019    1    2               11     1           Y  2019    1       2   

   ...  다태아분류코드  다태아출산순위코드  출생아체중  모총출생아수코드  모생존아수코드  부모동거기간  부_국적구분코드  \
0  ...        1          1  0.453         2       99     999         1   
1  ...        9          9  3.090        99       99     999         9   
2  ...        1          1  3.380         2        2       4         1   
3  ...        1          1  3.060         2        2       4         1   
4  ...        2          1  2.645         1        1       1         1   

   부_국적코드  모_국적구분코드  모_국적코드  
0     NaN         3  9000.0  
1     NaN         1     NaN  
2     NaN         1     NaN  
3     NaN         1     NaN  
4     NaN         1     NaN  

[5 rows x 29 columns]

항목명  코드 코드의미 및 단위
0  출생자주소지_행정구역시도코드  11     서울특별시
1              NaN  21     부산광역시
2              NaN  22     대구광역시
3              NaN  23     인천광역시
4              NaN  24     광주광역시

### 기계학습1
아래를 기준으로 데이터 전처리를 진행하라      
1. 년도별 출생 데이터를 하나의 데이터 프레임으로 합친다.     
2. 데이터 컬럼 상세 데이터를 바탕으로 출생 데이터의 값을 변환한다.
3. 변환한 데이터 컬럼 중 '미상'을 포함하는 행이 존재한다면 해당 행은 제거한다.    
4. 아래의 조건에 해당하는 각 컬럼의 값이 있다면 해당 행은 제거한다
    - 부_교육정도코드 : 학력없음    
    - 부_각세연령 : 999.0    
    - 모_교육정도코드 : 학력없음
    - 모_각세연령 :999.0
    - 실제결혼연도 : 9999.0
    - 실제결혼월 : 99.0
    - 부모동거기간 : 999
    - 모_각세연령 : 50세 이상
    - 임신주수 : 0
    - 출생아체중: 1kg 이하

결과예시
<img src="./data/1.png" style="width:10000px; float:left; ">

### 기계학습2
출생아체중을 중점으로 EDA를 실시하라 (시각화 포함)

### 기계학습3     
아래의 기준으로 추가적인 전처리를 실행하고 `출생아체중`이 각 컬럼들의 값에 따른 차이가 존재하는지 통계 검정을 실시하라
- 데이터 필터 : `신고월`이 2월이며 `신고일`이 10일 이하
- 컬럼제거 : 연도,신고연도,신고월,신고일, 출생자주소지_행정구역시도코드, 결혼중외의자녀여부코드,부_교육정도코드,모_직업분류코드,출생연도,출생월,출생장소코드,실제결혼연도,실제결혼월,모생존아수코드,부_국적코드,모_국적코드,모_국적구분코드,부_국적구분코드

### 기계학습4     
3에서 전처리한 데이터를 바탕으로 출생아 체중을 예측하는 회귀 모델을 만들려고한다. 아래의 조건에 맞게 모델링을 진행하고 평가하라
- 성별코드 기준으로 층화추출 진행 (학습데이터 70%, 평가 데이터 30%)
- 평가기준 rmse
- 랜덤포레스트, 다중선형회귀 모델로 진행

### 서울 지하철 일별 승하차 인원 데이터
출처 :https://data.seoul.go.kr/dataList/OA-12252/S/1/datasetView.do 후처리     
data_path = './data/subway.csv'


In [143]:
import pandas as pd
df = pd.read_csv('./data/subway.csv')
df.head()

날짜  역번호        역명  구분  05~06  06~07  07~08  08~09  09~10  10~11  \
0  2016-01-01  150  서울역(150)  승차    469    339    584   1144   1723   2087   
1  2016-01-01  150  서울역(150)  하차    342   1637   1753   1856   2438   2425   
2  2016-01-01  151   시청(151)  승차    103    127    110    191    298    582   
3  2016-01-01  151   시청(151)  하차     68    306    554    939    689    665   
4  2016-01-01  152   종각(152)  승차    702    341    234    296    335    419   

   11~12  12~13  13~14  14~15  15~16  16~17  17~18  18~19  19~20  20~21  \
0   2983   3527   3386   2612   3639   3554   3200   3076   2519   2281   
1   2548   2718   2973   2921   3341   3108   2656   2134   1859   1394   
2    700    604    731    900   1069   1339   1458   1499   1383   1284   
3    884    956   1084   1216   1033   1035    951    792    618    402   
4    733    865    991   1300   1772   2175   2580   2627   2302   2315   

   21~22  22~23  23~24  24~  호선  
0   2398   2056    996   58   1  
1   1344   1056    612  222   1  
2   1198    675    375    9   1  
3    309    192    114   34   1  
4   2365   2140   1117   51   1

### 기계학습5    
아래의 기준으로 전처리를 진행하라     
1. 일자마다 노선별로 각 시간대의 모든역에서의 이용 승객 수를 승차,하차 각각 더한다.
    - 날짜,호선,구분,모든시간대 컬럼만 남아있어야한다. (4386 rows × 23 columns)
2. 아래의 조건으로 새로운 데이터프레임을 생성한다. 일자마다 호선별로 각 시간대의 승차 / 하차 비율을 구하여 _ratio 를 붙혀 새로운 컬럼을 만든다.
    - 2016-01-01, 1호선의 `05 ~ 06(승차) /05 ~ 06(하차)` : 
        - 컬럼 : `05~06_ratio` 
        - 값 : 3.5696 (2613/732)
3. 해당 일자가 평일(0)인지 주말(1)인지 구분하는 'weekend' 컬럼을 추가한다

<img src='./data/2.png'>

### 기계학습6     
데이터 EDA를 진행하라. (시각화 포함)

### 기계학습7     
호선을 예측하는 분류 모델을 만들고 평가하라
- 평가데이터 : 일(day)이 4로 나눈 나머지가 1인 일자
- 학습데이터 : 그 외 
- 평가지표 : f1_score (micro,macro 모두 확인)
- 2가지 이상 모델

# 통계 (40점)

### 통계1      
한 카페에서는 고객들이 음료를 주문한 후, 음료가 준비되는 시간을 추적한다.      
음료 준비 시간은 평균이 5분이고, 표준편차가 1.5분인 정규분포를 따른다.         

1) 특정 고객이 음료를 6분 이내에 받을 확률은?     
2) 3분에서 7분 사이에 음료가 준비될 확률은?   
3) 10명의 고객 중, 6분 이내에 음료를 받을 고객이 3명 이상일 확률은?

### 통계2
한 레스토랑에서 두 명의 셰프(A와 B)가 준비한 요리에 대한 고객들의 만족도를 조사했다.     
각 셰프가 준비한 요리에 대해 10명의 고객이 점수를 매겼다. 점수는 1점부터 10점까지이며, 점수가 높을수록 만족도가 높음을 의미한다.     
두 셰프의 요리에 대한 고객 점수가 정규분포를 따르지 않는다고 가정하고, 셰프 A와 셰프 B의 요리에 대해 고객들의 만족도가 차이가 나는지 비모수 검정을 통해 확인하라.     

셰프 A의 고객 점수: 
7,8,9,5,6,7,8,6,7,9

셰프 B의 고객 점수: 
6,5,8,7,6,5,6,4,7,5

(1) 가설설정     
(2) 통계검정

### 통계3       
병원에서 특정 질병의 진단을 위해 두 가지 서로 다른 검사 (검사 X, 검사 Y)를 시행한다. 검사 X와 Y는 서로 독립적이며 각각 다른 진단 결과를 제공한다.    


- 전체 인구에서 질병을 가진 사람의 비율: 2% 
- 질병이 없는 사람의 비율: 98% 
- 질병이 있는 경우 검사 X가 양성일 확률: 73% 
- 질병이 없는 경우 검사 X가 양성일 확률: 10% 
- 질병이 있는 경우 검사 Y가 양성일 확률: 66% 
- 질병이 없는 경우 검사 Y가 양성일 확률: 5% 


1. 검사 X와 검사 Y 모두 양성일 때, 환자가 실제로 질병을 가질 확률은 얼마인가?
2. 검사 X가 양성이고 검사 Y가 음성일 때, 환자가 실제로 질병을 가질 확률은 얼마인가?
3. 검사 Y가 양성일 때, 검사 X가 음성일 확률은?